hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -239,10 +240,52 @@ def __stoppable_test_function(self, data):
                     # correct engine.
                     raise
 
-    def ir_tree_to_data(self, ir_tree_nodes):
+    def _cache(self, data):
+        result = data.as_result()
+        # when we shrink, we try out of bounds things, which can lead to the same
+        # data.buffer having multiple outcomes. eg data.buffer=b'' is Status.OVERRUN
+        # in normal circumstances, but a data with
+        # ir_nodes=[integer -5 {min_value: 0, max_value: 10}] will also have
+        # data.buffer=b'' but will be Status.INVALID instead.
+        #
+        # I think this indicates that we should be mark_overrun internally in
+        # these cases instead? alternatively, we'll map Status.INVALID to Overrun
+        # when caching.
+        if data.invalid_at is not None:
+            assert data.status is Status.INVALID
+            result = Overrun
+        self.__data_cache[data.buffer] = result
+        self.__data_cache_ir[tuple(data.examples.ir_tree_nodes)] = result
+
+    def cached_test_function_ir(self, ir_tree_nodes):
+        key = tuple(ir_tree_nodes)
+        try:
+            return self.__data_cache_ir[key]
+        except KeyError:
+            pass
+
+        try:
+            trial_data = ConjectureData.for_ir_tree(ir_tree_nodes)
+            self.tree.simulate_test_function(trial_data)
+        except PreviouslyUnseenBehaviour:
+            pass
+        else:
+            trial_data.freeze()
+            try:
+                return self.__data_cache_ir[tuple(trial_data.examples.ir_tree_nodes)]
+            except KeyError:
+                pass
+
         data = ConjectureData.for_ir_tree(ir_tree_nodes)
-        self.__stoppable_test_function(data)
-        return data
+        self.test_function(data)
+        self._cache(data)
+        # This covers slightly different cases than caching `data`. If the ir
+        # nodes (1) are not in our cache, (2) are not in our DataTree, and (3)
+        # running `data` through the test function concludes before consuming
+        # all ir nodes (eg because of misaligned ir nodes), the initial ir nodes
+        # will not be cached and we may continue to miss them.
+        self.__data_cache_ir[key] = data.as_result()
+        return data.as_result()
```

Comment:
ah, this isn't quite right. `DataTree` *should* be normalizing these buffers; `[0, 2, 0, 0]` where a `mark_invalid` happens between `2` and `0` should result in simulating any tree with a prefix of `[0, 2]` as also invalid. It doesn't right now because I've forgotten to add observers to `for_ir_tree`-based datas. 

But it's not that simple, because `[0, 2, n]` is only invalid if the test function requests `x != n`, otherwise the tree is actually aligned and valid. Probably we need native datatree support for invalid-because-misaligned conclusions.

Reply:
I've improved this in f7d0d29b6fa1a5a8e81afc8d5de07cbe080f6232. I think this is probably a good candidate for extracting to a separate pull.

Reply:
spent some more time on this and split the result to #3977.